### PR TITLE
feat: Prepare frontend to recieve database content

### PR DESCRIPTION
**Pull request by FlynnOConnell**
Capture

- Updated Cargo.toml for appropriately-cased module.
- Updated readme to reflect similar information as backend/readme. 
- Added header nav-bar with currently empty links.
- Set up a content area where we can route database requests to display information. 
- Added route navigation
- Restructure into components -> views
- Add deps for json, http requests
- Add component for an input form 
- Add component for each input text area of that form

The input-text area will accept a callback `onsubmit` that will do most of the heavy lifting.

We can test the response [here](https://reqres.in/), and this callback (I think) is where we will update the db from.

### CODE REVIEW

File: frontend/src/views/home.rs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+// use serde_json;
+// use wasm_bindgen::JsValue;
+// use web_sys::HtmlInputElement;
+
+use yew::prelude::*;
+
+use crate::components::input::InputField;
+
+#[derive(Clone, PartialEq, Properties, Debug, Default, Serialize, Deserialize)]
+pub struct RegistrationForm {
```
Comment (FlynnOConnell) [inline frontend/src/views/home.rs:12]: These types of structs are what I'd imagine would go in "common"

Comment (FlynnOConnell) [reply frontend/src/views/home.rs:12]: We can add database migrations to set up whatever type of data we want in here. I think we should shoot for something a little more fun but this is a starter.

File: frontend/src/views/home.rs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+                <InputField input_node_ref={last_name_ref} label={"Last Name".to_owned()} name={"last_name".clone()} field_type={"text".clone()}  />
+                <InputField input_node_ref={email_ref} label={"Email".to_owned()} name={"email".clone()} field_type={"email".clone()}  />
+                <InputField input_node_ref={password_ref} label={"Password".to_owned()} name={"password".clone()} field_type={"password".clone()}  />
+                <InputField input_node_ref={confirm_password_ref} label={"Confirm Password".to_owned()} name={"confirm_password".clone()} field_type={"password".clone()}  />
+                <p class="error-text">{ if *password_is_valid { "" } else { "Passwords do not match" } }</p>
+                <button type="submit" class="button button-primary">{"Submit"}</button>
+            </form>
+        </main>
+    }
+}
```
Comment (FlynnOConnell) [inline frontend/src/views/home.rs:52]: This is set up initially as a one-way street. I think getting a concrete db setup that we can have a persistent state in would be helpful. So if we set a few "users" or "messages", they will still be around after the db is shut down.

File: frontend/Cargo.toml [hunk 3]
```diff
@@ -9,10 +9,8 @@ license = "MIT OR Apache-2.0"
 keywords = ["yew", "trunk"]
 categories = ["gui", "wasm", "web-programming"]
 
-# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 [dependencies]
 yew = { version="0.20", features=["csr"] }
 yew-router = "0.15.0"
-yew_styles = { version="0.11", features=["layouts", "button"] }
```
Comment (FlynnOConnell) [inline frontend/Cargo.toml:16]: This turned out to be a pretty poor library. I did [find this](https://github.com/yewstack/yewtify) though, which is very similar to the framework I used for my website. Probably something I'll play around with. It's still pretty young.

**Comment by NeuroPyPy** [comment]
> * Updated Cargo.toml for appropriately-cased module.
> 
>     * Updated readme to reflect similar information as backend/readme.
> 
>     * Added header nav-bar with currently empty links.
> 
>     * Set up a content area where we can route database requests to display information.
> 
>     * Added route navigation
> 
>     * Restructure into components -> views
> 
>     * Add deps for json, http requests
> 
>     * Add component for an input form
> 
>     * Add component for each input text area of that form
> 
> 
> The input-text area will accept a callback `onsubmit` that will do most of the heavy lifting.
> 
> We can test the response [here](https://reqres.in/), and this callback (I think) is where we will update the db from.

File: frontend/Cargo.toml [hunk 4]
```diff
@@ -1,18 +1,22 @@
 [package]
-name = "trunk-template"
+name = "flunk_website"
 version = "0.1.0"
 edition = "2021"
-description = "Template for starting a Yew project using Trunk"
+description = "Frontend for Flunk Website"
 readme = "README.md"
-repository = "https://github.com/yewstack/yew-trunk-minimal-template"
+repository = "https://github.com/LukeBrandon/fluke"
```
Comment (FlynnOConnell) [inline frontend/Cargo.toml:7]: Not sure if this should be /frontend

File: frontend/Cargo.toml [hunk 5]
```diff
@@ -1,18 +1,22 @@
...
-yew-router = "0.15.0"
-yew_styles = { version="0.11", features=["layouts", "button"] }
-log = "0.4"
+yew-router = "0.17"
 wasm-logger = "0.2"
+wasm-bindgen = "0.2.84"
+web-sys = "0.3.61"
+wasm-bindgen-futures = "0.4"
+serde = "1.0.152"
+serde_json = "1.0.94"
```
Comment (FlynnOConnell) [inline frontend/Cargo.toml:21]: We already define these in common, I'm not sure if we need them here or if we just abolish the common folder entirely. I've seen a few repos use it

File: frontend/README.md [hunk 6]
```diff
@@ -24,51 +25,19 @@ Simply run the following command to install it:
...
-Finally, you should update this very `README` file to be about your app.
-
-### License
-
-The template ships with both the Apache and MIT license.
-If you don't want to have your app dual licensed, just remove one (or both) of the files and update the `license` field in `Cargo.toml`.
-
-There are two empty spaces in the MIT license you need to fill out: `` and `Luke Brandon <lukebrandon68@gmail.com>`.
+On WSL2 you may need to add some env variables to connect with ssl: 
+- `export OPENSSL_INCLUDE_DIR=/usr/include/openssl`
```
Comment (FlynnOConnell) [inline frontend/README.md:41]: These were the bane of my existence

File: frontend/index.html [hunk 7]
```diff
@@ -6,7 +6,7 @@
     <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/all.min.css">
     <link data-trunk rel="rust" />
     <link data-trunk rel="css" href="./index.css" />
-    <link data-trunk rel="copy-file" href="assets/logo_transparent.png"/>
+    <link data-trunk rel="copy-file" href="assets/twitter_header_photo_1.png"/>
```
Comment (FlynnOConnell) [inline frontend/index.html:9]: Need to make this transparent

File: frontend/src/app.rs [hunk 8]
```diff
@@ -1,11 +1,29 @@
...
+    Home,
+    #[not_found]
+    #[at("/404")]
+    NotFound,
+}
+
+pub fn switch(routes: Route) -> Html {
+    match routes {
+        Route::Home => html! { <Home /> },
+        Route::NotFound => html! { <h1>{ "404" }</h1> },
```
Comment (FlynnOConnell) [inline frontend/src/app.rs:18]: After form submission, add another route here to the poem. It was complaining when I tried to do this and its bedtime

File: frontend/src/components/landing.rs [hunk 9]
```diff
@@ -0,0 +1,83 @@
+use yew::prelude::*;
```
Comment (FlynnOConnell) [inline frontend/src/components/landing.rs:1]: Probably going to slash this file entirely but its helpful for reference

File: frontend/src/components/landing.rs [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+                <div>
+                   <PoemComponent />
+                </div>
+            </div>
+        </main>
+    }
+}
+
+#[function_component]
+fn PoemComponent() -> Html {
```
Comment (FlynnOConnell) [inline frontend/src/components/landing.rs:44]: Obviously not slashing this

File: frontend/Cargo.toml [hunk 11]
```diff
@@ -1,18 +1,22 @@
 [package]
-name = "trunk-template"
+name = "flunk_website"
```
Comment (LukeBrandon) [inline frontend/Cargo.toml]: bruh FLUNK? FLUKE!

File: frontend/Cargo.toml (hunk 4 above)
Comment (LukeBrandon) [reply frontend/Cargo.toml:7]: The repository is what is here, if we wanted to do submodules we could but that's just more work for now.

File: frontend/src/components/input.rs [hunk 12]
```diff
@@ -0,0 +1,36 @@
+use yew::prelude::*;
+
+#[derive(Clone, PartialEq, Properties)]
+pub struct InputFieldProps {
+    // pub input_value: String,
+    // pub on_cautious_change: Callback<ChangeData>,
```
Comment (LukeBrandon) [inline frontend/src/components/input.rs:6]:
```suggestion
```
?

Comment (LukeBrandon) [reply frontend/src/components/input.rs:6]: why are these commented out?

File: frontend/src/views/home.rs (hunk 2 above)
Comment (LukeBrandon) [reply frontend/src/views/home.rs:52]: Yeah getting a user set up would be good.  Then we have to start messing with authentication on the BE.  Unknown how ergonomic that is with Rocket, may be really easy

**Review by LukeBrandon — APPROVED** [review]
There is dead code, but good to keep things moving.  If there are things that are broken you can obviously fix them before mergind

File: frontend/src/views/home.rs [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+            <form class="registration-form">
+                <InputField input_node_ref={username} label={"Username".to_owned()} name={"username".clone()} field_type={"text".clone()} />
+                <InputField input_node_ref={first_name_ref} label={"First Name".to_owned()} name={"first_name".clone()} field_type={"text".clone()}  />
+                <InputField input_node_ref={last_name_ref} label={"Last Name".to_owned()} name={"last_name".clone()} field_type={"text".clone()}  />
+                <InputField input_node_ref={email_ref} label={"Email".to_owned()} name={"email".clone()} field_type={"email".clone()}  />
+                <InputField input_node_ref={password_ref} label={"Password".to_owned()} name={"password".clone()} field_type={"password".clone()}  />
+                <InputField input_node_ref={confirm_password_ref} label={"Confirm Password".to_owned()} name={"confirm_password".clone()} field_type={"password".clone()}  />
+                <p class="error-text">{ if *password_is_valid { "" } else { "Passwords do not match" } }</p>
+                <button type="submit" class="button button-primary">{"Submit"}</button>
+            </form>
```
Comment (LukeBrandon) [inline frontend/src/views/home.rs:49]: do us all a favor here and wrap this in a div with the css styles as follows
```
display: flex;
flex-direction: column;
row-gap: 8px;
```

File: frontend/src/components/input.rs (hunk 12 above)
Comment (FlynnOConnell) [reply frontend/src/components/input.rs:6]: They are causing some weird errors in <Input Field> components and I havent figured out the implementation of on_cautious_change yet.

**Review by FlynnOConnell — COMMENTED** [review]
There is a lot of dead code, but code I want to keep to implement. Fixed some of the smaller issues, and leaving the formatting of the input form, I have an idea of how I'd like it to look.

**Pull request merged**